internal/pkg/comm/creds.go

```diff
@@ -56,6 +58,9 @@ func (sc *serverCreds) ClientHandshake(context.Context,
 
 // ServerHandshake does the authentication handshake for servers.
 func (sc *serverCreds) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	sc.lock.Lock()
```

Comment:
TLS handshakes involve network I/O so we don't want to hold locks for a long time. Even if there is nothing competing this lock, it is usually considered bad practice to hold locks while doing network I/O.

I suggest that the TLS config will be cloned under a read lock and nothing more.